source/utils/numeric.cpp

```diff
@@ -0,0 +1,64 @@
+// Copyright (c) 2024 xlnt-community
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE
+//
+// @license: http://www.opensource.org/licenses/mit-license.php
+// @author: see AUTHORS file
+
+#include <xlnt/utils/numeric.hpp>
+#include <detail/locale.hpp>
+#include <detail/parsers.hpp>
+
+#include <iomanip>
+
+constexpr auto serialization_digits_full_precision = 15;
+constexpr auto serialization_digits_short = 6;
+
+std::string xlnt::detail::number_serialiser::serialise(double d) const
+{
+    std::ostringstream ss;
+    ss.imbue(detail::get_serialization_locale());
+    ss << std::defaultfloat << std::setprecision(serialization_digits_full_precision) << d;
+    return ss.str();
+}
+
+std::string xlnt::detail::number_serialiser::serialise_short(double d) const
+{
+    std::ostringstream ss;
+    ss.imbue(detail::get_serialization_locale());
+    ss << std::fixed << std::setprecision(serialization_digits_short) << d;
+    return ss.str();
+}
+
+double xlnt::detail::number_serialiser::deserialise(const std::string &s, ptrdiff_t *len_converted) const
+{
+    assert(!s.empty());
+    assert(len_converted != nullptr);
+    *len_converted = 0;
+    double d = std::numeric_limits<double>::quiet_NaN();
+    std::size_t num_parsed_characters = 0;
+    if (detail::parse(s, d, &num_parsed_characters))
+    {
+        // Prevent unwanted overflows.
+        assert(num_parsed_characters <= std::numeric_limits<ptrdiff_t>::max());
```

Comment:
Was it a mistake to use the type `ptrdiff_t` for `len_converted`? Maybe we can just change the type to `size_t` as this function is in the detail namespace.

Reply:
No, but... let me explain. I implemented the functions that use C++ strings very similarly to the way the C++ functions `std::stoi`, `std::stol` and `std::stoll` [have been standardized](https://en.cppreference.com/w/cpp/string/basic_string/stol) - that is, by using `size_t` for the number of characters. The thing is, `std::stoll` must call `std::strtoll` internally as specified by the C++ standard, which will return a pointer, so to calculate the number of characters, `std::stoll` will have to subtract the end pointer from the beginning pointer (resulting in `ptrdiff_t`), then cast the result to `size_t`. Subtracting pointers results in a signed `ptrdiff_t` because you could, of course, also do it the other way around, getting a negative result, so it makes sense that the result of subtracting pointers is a signed integer. However, in our case, it makes no sense to have a negative result, so while `ptrdiff_t` is technically more correct, I think that `size_t` is nicer to work with, as it shows that the result can never be negative. In this case, I personally agree to the C++ standard, which is why I designed the new parsing functions the same way.

I will change the type to `size_t`, like you proposed :wink: